### DISCUSSION TITLE
Constants / Root variables as Ops?

### DISCUSSION THREAD

**Discussion by ricardoV94** [pymc-devs/pytensor#924] [Ideas]
Most rewrites we have to do something like `if x.inputs[0].owner is not None and isinstance(x.inputs[0].owner.op, log)` , when we are looking for say a log operation. The first check is to rule out root variables / constants that have no owner and therefore no owner.op.

What if constants came out of a no-input Operation (parametrized by the constant value, so they can be distinguished merged), and root variables from a no-input Operation that can never be merged? Then we could always and safely do `isinstance(x.inputs[0].owner.op, log)` because all variables always have an owner and and Op (they may just not have any further inputs).

We did something similar recently for the Outputs of an OpFromGraph: https://github.com/pymc-devs/pytensor/pull/790

**Comment by ricardoV94** [comment]
Alternatively we can have a helper function `output_of(variable, type)` that takes care of the annoying `if owner is not None` part, but adds a callable overhead in our rewrites

**Comment by jessegrabowski** [comment]
I prefer the `Root` pseudo-`Op` option. It's elegant, and makes all the logic for rewrites consistent end-to-end. I agree the `if owner is None` checks everywhere are horrible.
